### PR TITLE
Allow visualizing and comparing nested params

### LINKED ISSUE #2838: [FR] Better visualization of nested params
## Proposal Summary

In my work, I compare runs with large and **nested** configurations stored as python dictionaries. This results in the Run comparison page looking like this:
mlflow_comparison_nested_configs

As shown, the parameters that are themselves dictionaries are displayed in a single row of the table. I'm guessing this is because they are retrieved as plain strings when the page is loaded.

The visualization would be much more useful if each level of a nested configuration was displayed indented and in a new row, much like in a .yaml file. Maybe this can be done with a (recurrent) check if a param can be read as dictionary.

In case of deep nesting, there could be a fixed upper limit (e.g. 3) supported by mlflow.
Ideally there would be a button in the UI, so the user can control the depth of nested params they want to see by expanding/contracting these params.

Moreover, since nested are not read as dictionaries, they are not searchable in order to filter runs.  

### What component(s), interfaces, languages, and integrations does this feature affect?
Components 
- [x] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interfaces
- [x] `area/uiux`: Front-end, user experience, JavaScript, plotting

### PR DESCRIPTION
**Pull request by jescalada**
<details><summary>&#x1F6E0 DevTools &#x1F6E0</summary>
<p>

[Open in GitHub Codespaces](https://codespaces.new/jescalada/mlflow/pull/13012?quickstart=1)

#### Install mlflow from this PR

```
pip install git+https://github.com/mlflow/mlflow.git@refs/pull/13012/merge
```

#### Checkout with GitHub CLI

```
gh pr checkout 13012
```

</p>
</details>

### Related Issues/PRs

Resolve #2838 

### What changes are proposed in this pull request?

In this PR, I added support for nested params (which would be logged directly as Python dictionaries). These are converted into JSON and then spread out in nested tables.

### How is this PR tested?

- [ ] Existing unit/integration tests
- [ ] New unit/integration tests
- [x] Manual tests

image

### Does this PR require documentation update?

- [x] No. You can skip the rest of this section.
- [ ] Yes. I've updated:
  - [ ] Examples
  - [ ] API references
  - [ ] Instructions

#### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [x] Yes. Give a description of this change to be included in the release notes for MLflow users.

Added support for visualizing nested params (when logging Python dictionaries directly for model configuration, etc.).

#### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/deployments`: MLflow Deployments client APIs, server, and third-party Deployments integrations
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface

- [x] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>

#### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [x] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

#### Should this PR be included in the next patch release?

`Yes` should be selected for bug fixes, documentation updates, and other small changes. `No` should be selected for new features and larger changes. If you're unsure about the release classification of this PR, leave this unchecked to let the maintainers decide.

<details>
<summary>What is a minor/patch release?</summary>

- Minor release: a release that increments the second part of the version number (e.g., 1.2.0 -> 1.3.0).
  Bug fixes, doc updates and new features usually go into minor releases.
- Patch release: a release that increments the third part of the version number (e.g., 1.2.0 -> 1.2.1).
  Bug fixes and doc updates usually go into patch releases.

</details>

- [x] Yes (this PR will be cherry-picked and included in the next patch release)
- [ ] No (this PR will be included in the next minor release)

### CODE REVIEW

**Comment by jescalada** [comment]
Pinging @daniellok-db for visibility... Thank you! :)

**Comment by daniellok-db** [comment]
Oops, i must have missed this one, sorry! I'll get to reviewing soon

File: mlflow/server/js/src/experiment-tracking/components/CompareRunView.tsx [hunk 1]
```diff
@@ -703,4 +749,20 @@ const mapStateToProps = (state: any, ownProps: any) => {
...
+  try {
+    const jsonString = value.replace(/'/g, '"');
+    return JSON.parse(jsonString);
+  } catch (e) {
+    console.error('Failed to parse string to JSON:', e);
+    return null;
+  }
+};
+
+// @ts-expect-error TS(2769): No overload matches this call.
```
Comment (daniellok-db) [inline mlflow/server/js/src/experiment-tracking/components/CompareRunView.tsx:767]: i think this line can be removed
```suggestion
```

File: mlflow/server/js/src/experiment-tracking/components/CompareRunView.css [hunk 2]
```diff
@@ -67,3 +67,26 @@
...
+}
+
+.json-table td:nth-child(2) {
+  width: 80%; /* Value cells take up the rest of the space */
+  white-space: pre-wrap;
+  word-wrap: break-word;
+}
+.truncate-text.single-line {
+  display: inline-block;
+}
```
Comment (daniellok-db) [inline mlflow/server/js/src/experiment-tracking/components/CompareRunView.css:92]: newline
```suggestion
}

```

**Review by daniellok-db — COMMENTED** [review]
i agree with having a solution for visualizing nested dicts, but i wonder if it might be confusing for users to see a table. i wonder if it's more conventional to just have pretty-printed JSON instead.

cc @hubertzub-db any thoughts?

File: mlflow/server/js/src/experiment-tracking/components/CompareRunView.tsx (hunk 1 above)
Comment (jescalada) [reply mlflow/server/js/src/experiment-tracking/components/CompareRunView.tsx:767]: It seems we get an error when removing that line:

image

Fortunately, I think I managed to fix it when [refactoring the component](https://github.com/mlflow/mlflow/pull/13288)!

**Comment by jescalada** [comment]
@daniellok-db I implemented the pretty-printed version and it honestly looks better than the tabular format (not to mention that the code is a lot more readable)!

**Review by daniellok-db — APPROVED** [review]
lgtm!

**Comment by daniellok-db** [comment]
merging as the test failure is unrelated

**Comment by ohmatheus** [comment]
cannot see same result on 3.1.4

**Pull request merged**